bridge-ffi/src/lib.rs

```diff
@@ -362,23 +444,393 @@ pub extern "C" fn tmprl_poll_activity_task(
+) {
+    let core = unsafe { &mut *core };
+    let req = match tmprl_core_t::decode_proto::<bridge::CompleteActivityTaskRequest>(
+        req_proto,
+        req_proto_len,
+    ) {
+        Ok(req) => req,
+        Err(message) => {
+            let resp = bridge::CompleteActivityTaskResponse {
+                error: Some(bridge::complete_activity_task_response::Error { message }),
+            };
+            unsafe {
+                callback(user_data, core.encode_proto(&resp).into_raw());
+            }
+            return;
+        }
+    };
+    let user_data = UserDataHandle(user_data);
+    core.tokio_runtime.clone().spawn(async move {
+        match core.complete_activity_task(req).await {
+            Ok(()) => unsafe {
+                callback(
+                    user_data.into(),
+                    &*DEFAULT_COMPLETE_ACTIVITY_TASK_RESPONSE_BYTES,
+                );
+            },
+            Err(err) => {
+                let resp = bridge::CompleteActivityTaskResponse { error: Some(err) };
+                unsafe { callback(user_data.into(), core.encode_proto(&resp).into_raw()) };
+            }
+        }
+    });
+}
+
+/// Record activity heartbeat.
+///
+/// The req_proto and req_proto_len represent a byte array for a
+/// RecordActivityHeartbeatRequest protobuf message. The callback is invoked
+/// on completion with a RecordActivityHeartbeatResponse protobuf message.
+#[no_mangle]
+pub extern "C" fn tmprl_record_activity_heartbeat(
+    core: *mut tmprl_core_t,
+    req_proto: *const u8,
+    req_proto_len: libc::size_t,
+    user_data: *mut libc::c_void,
+    callback: tmprl_callback,
+) {
+    let core = unsafe { &mut *core };
+    let req = match tmprl_core_t::decode_proto::<bridge::RecordActivityHeartbeatRequest>(
+        req_proto,
+        req_proto_len,
+    ) {
+        Ok(req) => req,
+        Err(message) => {
+            let resp = bridge::RecordActivityHeartbeatResponse {
+                error: Some(bridge::record_activity_heartbeat_response::Error { message }),
+            };
+            unsafe {
+                callback(user_data, core.encode_proto(&resp).into_raw());
+            }
+            return;
+        }
+    };
+    let user_data = UserDataHandle(user_data);
+    // We intentionally spawn even though the core call is not async so the
+    // callback can be made in the tokio runtime
+    core.tokio_runtime.clone().spawn(async move {
+        core.record_activity_heartbeat(req);
+        unsafe {
+            callback(
+                user_data.into(),
+                &*DEFAULT_RECORD_ACTIVITY_HEARTBEAT_RESPONSE_BYTES,
+            );
+        }
+    });
+}
+
+/// Request workflow eviction.
+///
+/// The req_proto and req_proto_len represent a byte array for a
+/// RequestWorkflowEvictionRequest protobuf message. The callback is invoked
+/// on completion with a RequestWorkflowEvictionResponse protobuf message.
+#[no_mangle]
+pub extern "C" fn tmprl_request_workflow_eviction(
```

Comment:
Almost all these functions have very similar structure. I think it's worth seeing if we can de-dupe that structure with macro or a higher-order function.